### PR TITLE
WIP/ERR: raise consistent errors for non-existent files

### LINKED ISSUE #13872: ERR: set errno in read_csv
The same is true for Pandas on Python 2.7 with except IOError and on Python 3.5 with OSError. The following is how to reproduce the bug on Python 3.5. 

Pandas is not reporting any errno within the exceptions. As a result, there is no way to differentiate between File Not Found Error (ENOENT) and other types of OSErrors. 
#### Code Sample

```
try:
    pd.read_csv("dummy_file")
except OSError as e:
    print(e.errno, errno.ENOENT)
```
#### Current Output

(None, 2)
#### Expected Output

(2, 2)
## INSTALLED VERSIONS

commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Darwin
OS-release: 15.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: 23.0.0
Cython: 0.24
numpy: 1.11.1
scipy: 0.17.1
statsmodels: 0.6.1
xarray: None
IPython: 4.2.0
sphinx: 1.4.1
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: 1.1.0
tables: 3.2.2
numexpr: 2.6.0
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.2
lxml: 3.6.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.13
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.40.0
pandas_datareader: None

### LINKED ISSUE #17918: pandas.read_ functions crashes on read from url with space characters
```python
import pandas
pandas.read_json('http://httpbin.org/anything/something something')
```
python throws exception
`HTTPError: HTTP Error 505: HTTP Version Not Supported`

if url escaped:
```
pandas.read_json('http://httpbin.org/anything/something%20something')
```
then it works as expected

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-96-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: ru_RU.UTF-8
LANG: ru_RU.UTF-8
LOCALE: None.None

pandas: 0.20.3
pytest: None
pip: 9.0.1
setuptools: 36.5.0
Cython: None
numpy: 1.13.3
scipy: None
xarray: None
IPython: 5.5.0
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: 2.7.3.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #29125: Unify error messages when opening inexistent excel/csv file
Similar to https://github.com/pandas-dev/pandas/issues/29102, excel and csv file extensions now have inconsistent errors for nonexistent files.

```
import pandas as pd
pd.read_csv('file_1.csv')
IOError: [Errno 2] File file_1.csv does not exist: 'file_1.csv'

pd.read_excel('file_2.xlsx')
IOError: [Errno 2] No such file or directory: 'file_2.xlsx'
```

Would be better to have a uniform error message across multiple file extensions, like the one with json in https://github.com/pandas-dev/pandas/pull/29104:

```
FileNotFoundError: File 'no_file.json' does not exist
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #29125
- [x] closes #13872
- [ ] closes #17918  <-- needs tests and double-checking
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

xref #29102 the json case is xfailed here.

### CODE REVIEW

File: pandas/tests/io/test_common.py [hunk 1]
```diff
@@ -360,3 +361,65 @@ def test_unknown_engine(self):
...
+        pd.read_table,
+        pd.read_fwf,
+        pytest.param(pd.read_excel, marks=[td.skip_if_no("xlrd")]),
+        pytest.param(
+            pd.read_json,
+            marks=[
+                pytest.mark.xfail(
+                    reason=(
+                        "Needs to distinguish between a path to read "
+                        "vs a string to parse"
```
Comment (jbrockmendel) [inline pandas/tests/io/test_common.py]: @WillAyd any thoughts on this?  I think I saw a semi-recent discussion about trying to guess whether a string is supposed to be a path or actual content.

Comment (WillAyd) [reply pandas/tests/io/test_common.py]: Yea this was a discussion with @datapythonista and @gfyoung . I don't have a strong point of view in any direction but would be against jumping through a lot of hoops to disambiguate.

I guess if we cared to disambiguate could probably just look for `[`, `{`, `"` or any leading number (ignoring whitespace) to start as those are the requirements for JSON

**Review by jreback — CHANGES_REQUESTED** [review]
we have to be really careful about this
as this basically reinventing the wheel and this raises lots of security concerns

**Comment by jbrockmendel** [comment]
> reinventing the wheel

If there's an existing wheel we can use, I'm all for that.

Going through the issues, I'm going to try to round up all of the "should we accept strings too?" etc issues so we can at least discuss them all at once.

Also happy to break parts of this off if anything can be identified as "easy"

File: pandas/tests/io/test_common.py (hunk 1 above)
Comment (gfyoung) [reply pandas/tests/io/test_common.py]: I would leave the JSON disambiguation alone for now.  I'm still not really that inclined to change at the moment.

Comment (jbrockmendel) [reply pandas/tests/io/test_common.py]: Looking through old issues, there are a lot that revolve around disambiguating paths vs content, e.g. #5924 for pickle.  (one clever suggestion was to have `pd.read_pickles`, `pd.read_jsons`, etc as analogues to the stdlib `json.loads`)

This doesn't need to be resolved for this PR, but I think we're going to need to have a decision about whether/when/how to guess if a string is a path or not.

**Comment by jbrockmendel** [comment]
Closing, will re-implement in smaller pieces.

**Pull request closed without merging**